keras/engine/training.py

```diff
@@ -1053,49 +1089,77 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=32,
-                    if isinstance(ins[-1], float):
-                        # Do not slice the training phase flag.
-                        ins_batch = _slice_arrays(ins[:-1], batch_ids) + [ins[-1]]
-                    else:
-                        ins_batch = _slice_arrays(ins, batch_ids)
-                except TypeError:
-                    raise TypeError('TypeError while preparing batch. '
-                                    'If using HDF5 input data, '
-                                    'pass shuffle="batch".')
-                batch_logs = {}
-                batch_logs['batch'] = batch_index
-                batch_logs['size'] = len(batch_ids)
-                callbacks.on_batch_begin(batch_index, batch_logs)
-                outs = f(ins_batch)
-                if not isinstance(outs, list):
-                    outs = [outs]
-                for l, o in zip(out_labels, outs):
-                    batch_logs[l] = o
+            if steps_per_epoch is not None:
+                for step_num in range(steps_per_epoch):
+                    batch_logs = {}
+                    batch_logs['batch'] = step_num
+                    batch_logs['size'] = 1
+                    callbacks.on_batch_begin(step_num, batch_logs)
+                    outs = f(ins)
 
-                callbacks.on_batch_end(batch_index, batch_logs)
-                if callback_model.stop_training:
-                    break
+                    if not isinstance(outs, list):
+                        outs = [outs]
+                    for l, o in zip(out_labels, outs):
+                        batch_logs[l] = o
 
-                if batch_index == len(batches) - 1:  # Last batch.
-                    if do_validation:
-                        val_outs = self._test_loop(val_f, val_ins,
-                                                   batch_size=batch_size,
-                                                   verbose=0)
-                        if not isinstance(val_outs, list):
-                            val_outs = [val_outs]
-                        # Same labels assumed.
-                        for l, o in zip(out_labels, val_outs):
-                            epoch_logs['val_' + l] = o
+                    callbacks.on_batch_end(step_num, batch_logs)
+                    if callback_model.stop_training:
+                        break
+
+                if do_validation:
+                    val_outs = self._test_loop(val_f, val_ins,
+                                               batch_size=batch_size,
+                                               steps=validation_steps,
+                                               verbose=0)
+                    if not isinstance(val_outs, list):
+                        val_outs = [val_outs]
+                    # Same labels assumed.
+                    for l, o in zip(out_labels, val_outs):
+                        epoch_logs['val_' + l] = o
+            else:
+                if shuffle == 'batch':
```

Comment:
Maybe add a warning that if `steps_per_epoch is not None`, `shuffle=True` has no effect.  (no test required since pytest is flaky when it catches warning)